src/main/java/org/opensearch/security/privileges/dlsfls/AbstractRuleBasedPrivileges.java

```diff
@@ -0,0 +1,833 @@
+     * given resolved indices.
+     *
+     * @throws PrivilegesEvaluationException If something went wrong during privileges evaluation. In such cases, any
+     *                                       access should be denied to make sure that no unauthorized information is exposed.
+     */
+    public boolean isUnrestricted(PrivilegesEvaluationContext context, IndexResolverReplacer.Resolved resolved)
+        throws PrivilegesEvaluationException {
+        if (context.getMappedRoles().isEmpty()) {
+            return false;
+        }
+
+        if (this.dfmEmptyOverridesAll
+            && CollectionUtils.containsAny(this.staticRules.rolesWithIndexWildcardWithoutRule, context.getMappedRoles())) {
+            return true;
+        }
+
+        if (resolved == null) {
+            return false;
+        }
+
+        if (this.hasRestrictedRulesWithIndexWildcard(context)) {
+            return false;
+        }
+
+        StatefulRules<SingleRule> statefulRules = this.statefulRules;
+
+        // The logic is here a bit tricky: For each index/alias/data stream we assume restrictions until we found an unrestricted role.
+        // If we found an unrestricted role, we continue with the next index/alias/data stream. If we found a restricted role, we abort
+        // early and return true.
+
+        for (String index : resolved.getAllIndicesResolved(context.getClusterStateSupplier(), context.getIndexNameExpressionResolver())) {
+            if (this.dfmEmptyOverridesAll) {
+                // We assume that we have a restriction unless there are roles without restriction.
+                // Thus, we only have to check the roles without restriction.
+                if (!this.hasUnrestrictedRulesExplicit(context, statefulRules, index)) {
+                    return false;
+                }
+            } else {
+                // if dfmEmptyOverwritesAll == false, we prefer restricted roles over unrestricted ones.
+                // Thus, we first check for restricted roles. Only if there are not any restricted roles,
+                // we check for the presence of unrestricted roles. If there are not any matching roles,
+                // we also assume full restrictions.
+
+                if (this.hasRestrictedRulesExplicit(context, statefulRules, index)) {
+                    return false;
+                } else if (!CollectionUtils.containsAny(this.staticRules.rolesWithIndexWildcardWithoutRule, context.getMappedRoles())
+                    && !this.hasUnrestrictedRulesExplicit(context, statefulRules, index)) {
+                        return false;
+                    }
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns true if there are roles without a rule which imposes restrictions for the particular index.
+     * Does consider rules with index wildcards ("*").
+     */
+    public boolean isUnrestricted(PrivilegesEvaluationContext context, String index) throws PrivilegesEvaluationException {
```

Comment:
Is it possible to remove the throws declaration from here and any other associated methods?

Reply:
At the moment, these methods are used in `DlsFlsValveImpl`: 

https://github.com/opensearch-project/security/blob/1964b4d68abe5714040359ceb76833b097ac0b46/src/main/java/org/opensearch/security/configuration/DlsFlsValveImpl.java#L142-L150

In case we get an exception here, we fail fast and deny the request processing:

https://github.com/opensearch-project/security/blob/1964b4d68abe5714040359ceb76833b097ac0b46/src/main/java/org/opensearch/security/configuration/DlsFlsValveImpl.java#L339-L346

If we catch the exception inside the `isUnrestricted()` method, we would need to decide what `boolean` to return. `true` would be wrong, as this would then give the user full privileges - while they likely should have restricted privileges. If we return `false`, we would continue to the further DLS/FLS processing, which would likely throw the same or a similar exception as the one that was just caught.

So, we would double up the logged exceptions in the logs - possibly spread over several shards. For analyzing issues, I would argue that it is better to have a single exception in one place.

What do you think?

Reply:
Where is this exception actually thrown though? My IDE shows that it can be removed.

There is a place where its thrown within `hasRestrictedPatternTemplates` and `hasUnrestrictedPatternTemplates`, but its caught and logged as an error. Do you mean to throw it up the chain?

Reply:
One method was indeed catching and ignoring the exception, while it should not have done this. I have changed the method and added tests to cover the particular situation:

https://github.com/opensearch-project/security/pull/4380/commits/3203f038a852a1acb122afb3227670dc21e2b5d6